src/common/base/EnumArray.h

```diff
@@ -0,0 +1,249 @@
+                >::reType::value,
+    };
+};
+
+template<typename E>
+class DefaultTransformer {
+public:
+    static uint64_t toValue(E e) {
+        return static_cast<uint64_t>(e);
+    }
+
+    static E toEnum(uint8_t v) {
+        return static_cast<E>(v);
+    }
+};
+
+/**
+ * Compact enum array, the idea comes from bitset.
+ * If the enum has two values, it eauals bitset.
+ * */
+template<typename E,  // the type of enum
+         size_t N,  // total numbers of enums
+         // total values number of the E, by default it will read the _ITEMS_NUM field
+         uint8_t EnumNum = E::_ITEMS_NUM,
+         // Implement toIndex and toValue methods
+         typename Transformer = DefaultTransformer<E>,
+         typename = std::enable_if_t<std::is_enum<E>::value>>
+class EnumArray {
+public:
+    class EnumArrayIterator: public std::iterator<
+                                    std::input_iterator_tag,   // iterator_category
+                                    E,                         // value_type
+                                    size_t,                    // difference_type
+                                    E*,                        // pointer
+                                    E                          // reference
+                                                 > {
+    public:
+        explicit EnumArrayIterator(EnumArray<E, N, EnumNum, Transformer>* arr, size_t idx = 0)
+            : arr_(arr)
+            , idx_(idx) {
+            if (idx_ > N) {
+                idx_ = N;
+            } else if (idx_ < 0) {
+                idx_ = 0;
+            }
+        }
+
+        EnumArrayIterator& operator++() {
+            idx_++;
+            return *this;
+        }
+
+        EnumArrayIterator operator++(int) {
+            iterator retval = *this;
+            ++(*this);
+            return retval;
+        }
+
+        EnumArrayIterator operator+(size_t diff) {
+            return EnumArrayIterator(arr_, idx_ + diff);
+        }
+
+        EnumArrayIterator operator-(size_t diff) {
+            return EnumArrayIterator(arr_, idx_ - diff);
+        }
+
+        bool operator==(const EnumArrayIterator& other) const {
+            return idx_ == other.idx_;
+        }
+
+        bool operator!=(const EnumArrayIterator& other) const {
+            return !(*this == other);
+        }
+
+        E operator*() const {
+            return arr_->get(idx_);
+        }
+
+    private:
+        EnumArray<E, N, EnumNum, Transformer>* arr_ = nullptr;
+        size_t idx_ = 0;
+    };
+
+    typedef EnumArrayIterator iterator;
+
+    EnumArray() {
+        blocks_.fill(0);
+    }
+    ~EnumArray() = default;
+
+    void put(size_t idx, E e) {
+        auto v = Transformer::toValue(e);
+        DCHECK(idx < kTotalBits) << "idx " << idx << ", total " << kTotalBits;
+        auto bIdx = blockIndex(idx);
+        auto offset = bitOffset(idx);
+        uint64_t mask = v << offset;
+        blocks_[bIdx] |= mask;
```

Comment:
is need set 0 to blocks_[bIdx]? if call put with same idx but diffirent e, then maybe can not get correct e?

Reply:
Good point!  We need to handle it.